programs/helium-sub-daos/src/instructions/staking/stake_v0.rs

```diff
@@ -0,0 +1,246 @@
+  pub clockwork: Program<'info, ThreadProgram>,
+}
+
+pub fn handler(ctx: Context<StakeV0>, args: StakeArgsV0) -> Result<()> {
+  let sum: u8 = args.percentages.iter().sum();
+  assert!(sum == 100);
+
+  // load the vehnt information
+  let voter = ctx.accounts.vsr_voter.load()?;
+  let registrar = &ctx.accounts.registrar.load()?;
+  let d_entry = voter.deposits[args.deposit_entry_idx as usize];
+  let voting_mint_config = &registrar.voting_mints[d_entry.voting_mint_config_idx as usize];
+  let curr_ts = registrar.clock_unix_timestamp();
+  let available_vehnt = d_entry.voting_power(voting_mint_config, curr_ts)?;
+
+  let seconds_left = d_entry
+    .lockup
+    .seconds_left(curr_ts)
+    .checked_sub(10)
+    .unwrap();
+  let future_ts = curr_ts
+    .checked_add(seconds_left.try_into().unwrap())
+    .unwrap();
+  let future_vehnt = d_entry.voting_power(voting_mint_config, future_ts)?;
+
+  let fall_rate = calculate_fall_rate(available_vehnt, future_vehnt, seconds_left).unwrap();
+
+  let to_stake_vehnt_amount = std::cmp::min(args.vehnt_amount, available_vehnt);
+
+  let curr_epoch = current_epoch(ctx.accounts.clock.unix_timestamp);
+
+  // underlying_hnt = amount_deposited_native * vehnt_amount / available_vehnt
+  // position_fall_rate = fall_rate * vehnt_amount / available_vehnt
+  let underlying_hnt = PreciseNumber::new(d_entry.amount_deposited_native.into())
+    .unwrap()
+    .checked_mul(&PreciseNumber::new(to_stake_vehnt_amount.into()).unwrap())
+    .unwrap()
+    .checked_div(&PreciseNumber::new(available_vehnt.into()).unwrap())
+    .unwrap();
+  let position_fall_rate: u64 = u128::from(fall_rate)
+    .checked_mul(to_stake_vehnt_amount.into())
+    .unwrap()
+    .checked_div(available_vehnt.into())
+    .unwrap()
+    .try_into()
+    .ok()
+    .unwrap();
+
+  let sub_daos = &mut ctx.remaining_accounts.to_vec();
+  let stake_position = &mut ctx.accounts.stake_position;
+  assert!(sub_daos.len() == stake_position.allocations.len());
+
+  let mut old_position_vehnt = 0;
+
+  if stake_position.last_claimed_epoch > 0 {
+    // updating pos, assert that all available rewards have been claimed
+    assert!(stake_position.last_claimed_epoch >= curr_epoch - 1);
+
+    old_position_vehnt = calculate_voting_power(
+      d_entry,
+      voting_mint_config,
+      stake_position.hnt_amount,
+      stake_position.hnt_amount,
+      curr_ts,
+    )?;
+  }
+
+  // update all the subdao stakes
+  for (i, sd_acc_info) in sub_daos
+    .iter()
+    .enumerate()
+    .take(stake_position.allocations.len())
+  {
+    if (stake_position.allocations[i].percent == 0 && args.percentages[i] == 0)
+      || sd_acc_info.key() == Pubkey::default()
+    {
+      continue;
+    }
+
+    assert!(
+      stake_position.last_claimed_epoch == 0
+        || stake_position.allocations[i].sub_dao == sd_acc_info.key()
+    ); // must be new account or the subdao keys should be equal
+    assert!(sd_acc_info.is_writable);
```

Comment:
It seems like we're missing a check that the sub_dao is part of the right dao with the right hnt?

Reply:
Yep, added that check to all the relevant instructions. Also added a missing check on registrar seeds